src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -262,29 +269,24 @@ private void EnqueueWork(bool highPriority, bool frozenPartialSemantics, Cancell
                 // bails gracefully by checking below.
                 await _visibilityTracker.DelayWhileNonVisibleAsync(
                     _dataSource.ThreadingContext, _dataSource.AsyncListener, _subjectBuffer, DelayTimeSpan.NonFocus, cancellationToken).NoThrowAwaitable(captureContext: true);
-
-                if (cancellationToken.IsCancellationRequested)
-                    return default;
             }
 
-            _dataSource.ThreadingContext.ThrowIfNotOnUIThread();
-            if (cancellationToken.IsCancellationRequested)
-                return default;
-
             using (Logger.LogBlock(FunctionId.Tagger_TagSource_RecomputeTags, cancellationToken))
             {
+                _dataSource.ThreadingContext.ThrowIfNotOnUIThread();
+                if (cancellationToken.IsCancellationRequested)
+                    return default;
+
                 // Make a copy of all the data we need while we're on the foreground.  Then switch to a threadpool
                 // thread to do the computation. Finally, once new tags have been computed, then we update our state
                 // again on the foreground.
                 var spansToTag = GetSpansAndDocumentsToTag();
                 var caretPosition = _dataSource.GetCaretPoint(_textView, _subjectBuffer);
-                var oldTagTrees = this.CachedTagTrees;
-                var oldState = this.State;
 
-                var textChangeRange = this.AccumulatedTextChanges;
-                var subjectBufferVersion = _subjectBuffer.CurrentSnapshot.Version.VersionNumber;
-
-                await TaskScheduler.Default;
+                // If we're being called from within a blocking JTF.Run call, we don't want to switch to the background
+                // if we can avoid it.
+                if (!calledFromJtfRun)
+                    await TaskScheduler.Default;
```

Comment:
new code.  avoiding a BG switch in the blocking JTF case.

Reply:
yeah, i brought that up in the other PR.  I wasn't sure about these helpers.  For one, it's very complex and takes locks.  There is IsMainThreadBlockedMaybe which is much lighter weight.  I'm def worried about anything expensive in tagging, which is why i went for the easy approach of the caller just passing this in.

THoughts?

Reply:
going to hold off for now on this.  we have another pr on this topic where we can reviist this.